src/main/resources/org/jenkinsci/plugins/workflow/job/console/NewNodeConsoleNote/script.js

```diff
@@ -1,62 +1,78 @@
-Behaviour.specify("span.pipeline-new-node", 'NewNodeConsoleNote', 0, function(e) {
-    if (e.processedNewNodeConsoleNote) {
-        return
-    }
-    e.processedNewNodeConsoleNote = true
-    var label = e.getAttribute('label')
-    if (label != null) {
-        var html = e.innerHTML
-        var suffix = ' (' + label.escapeHTML() + ')';
-        if (html.indexOf(suffix)<0) {
-            e.innerHTML = e.innerHTML.replace(/.+/, '$&' + suffix) // insert before EOL
+Behaviour.specify("span.pipeline-new-node", 'NewNodeConsoleNote', 0, function (e) {
+    /**
+     * The setTimeout() method calls a function or evaluates an expression after a specified number of milliseconds.
+     * It is used here to prevent a browser freeze that might occur if a very big log is being evaluated.
+     */
+    setTimeout(() => {
+        var nodeId = e.getAttribute('nodeId')
+        var oid = e.getAttribute('nodeId')
+        var startId = e.getAttribute('startId')
+        var label = e.getAttribute('label')
+        var nodes = $$('.pipeline-new-node')
+        var enclosings = new Map() // id → enclosingId
+        var labels = new Map() // id → label
+
+        if (e.processedNewNodeConsoleNote) {
+            return
         }
-    }
-    var nodeId = e.getAttribute('nodeId')
-    var startId = e.getAttribute('startId')
-    if (startId == null || startId == nodeId) {
-        e.innerHTML = e.innerHTML.replace(/.+/, '$&<span class="pipeline-show-hide"> (<a href="#" onclick="showHidePipelineSection(this); return false">hide</a>)</span>')
-        // TODO automatically hide second and subsequent branches: namely, in case a node has the same parent as an earlier one
-    }
-    // The CSS rule for branch names only needs to be added once per node, so we
-    // check in case we are viewing the truncated log and have already processed
-    // a duplicate synthetic span element for this node.
-    var maybeDupeNodes = $$('[nodeid=\"'+nodeId+'\"].pipeline-new-node');
-    for (var i = 0; i < maybeDupeNodes.length; i++) {
-        var node = maybeDupeNodes[i];
-        if (node !== e && node.processedNewNodeConsoleNote) {
-            return;
+        e.processedNewNodeConsoleNote = true
+
+        if (label != null) {
+            var html = e.innerHTML
+            var suffix = ' (' + label.escapeHTML() + ')';
+            if (html.indexOf(suffix) < 0) {
+                e.innerHTML = e.innerHTML.replace(/.+/, '$&' + suffix) // insert before EOL
+            }
         }
-    }
-    var nodes = $$('.pipeline-new-node')
-    var enclosings = new Map() // id → enclosingId
-    var labels = new Map() // id → label
-    for (var i = 0; i < nodes.length; i++) {
-        var node = nodes[i]
-        var oid = node.getAttribute('nodeId')
-        enclosings.set(oid, node.getAttribute('enclosingId'))
-        labels.set(oid, node.getAttribute('label'))
-    }
```

Comment:
Removing this loop causes branch names to no longer be shown when log output changes from one branch to another in many cases.

The problem is that in the `while` loop right after this, we need to walk through all of this node's parents to find what parallel branch it is a part of. Because the loop is gone, now we only walk through the immediate parents of each node, so if there is any nesting inside of the parallel branch, for example because of a block-scope step, we never find the enclosing parallel branch, and so the `while` loop exits before finding a label.

The old code is not optimal because we create a fresh `enclosings` and `labels` map for every node even though most of the content will be the same. I would try to come up with a good way to store those as page-level state so that you can keep the approach you have now of only updating `enclosings` and `labels` once per node, and then split the `while` loop below out somehow and only run it after the `enclosings` and `labels` have data for every node (not sure of a good way to do that).

Here is a Pipeline you can run to compare the behavior with and without this PR to see what I mean:

```
stage('Outer') {
    parallel(branchOne: {
        stage('Stage One') {
            node() {
                sh 'echo "from one"'
                sleep(time: 1)
                echo 'from one again'
            }
        }
    }, branchTwo: {
        stage('Stage Two') {
            node() {
                sh 'echo "from two"'
                sleep(time: 1)
                echo 'from two again'
            }
        }
    })
}
```

Reply:
I see. I will try to find a solution to this problem.